### ISSUE TITLE
pathBuilder is not getting browser capabilities

### ISSUE THREAD

**Issue by mlison** [mlison/protractor-jasmine2-screenshot-reporter#4]
pathBuilder in specDone method doesn't get browser capabilities passed to it.

https://github.com/mlison/protractor-jasmine2-screenshot-reporter/blob/master/index.js#L143

(same applies for metadataBuilder)

**Comment by bryanlb** [comment]
+1

**Comment by mlison** [comment]
There probably won't be any good / proper solutions until this is resolved: https://github.com/angular/protractor/issues/1938

**Comment by eolognt** [comment]
Same here :+1:

**Comment by flaviomeira** [comment]
+1

**Comment by Priyapanch** [comment]
Same issue

**Comment by NickLargen** [comment]
What is the issue that prompted the [save filename before making a screenshot to avoid promise trouble](https://github.com/mlison/protractor-jasmine2-screenshot-reporter/commit/4e445e4f8fbf4512172a87883295953508c47e5e) commit? pathBuilder worked correctly for me with that change reverted.

**Comment by haugthom** [comment]
+1

**Comment by devmao** [comment]
\-        file = opts.pathBuilder(spec, suites);
+       file = opts.pathBuilder(spec, suites, spec._capabilities);

**Comment by michaschwab** [comment]
@devmao That didn't work for me, as spec doesn't seem to have that attribute - at least for me.

Thanks @NickLargen - undoing those changes fixes it for me, too.

**Comment by devmao** [comment]
@michaschwab Which version of protractor are you using? 
I tested it only with 2.4.0 and jasmine 2.3.2

**Comment by bvarga** [comment]
another workaround:

``` javascript
onPrepare: function() {
    browser.getCapabilities().then(function (capabilities) {
      browser.capabilities = capabilities;
    });

    jasmine.getEnv().addReporter(
      new HtmlScreenshotReporter({
        dest: 'test/results/screenshots',
        filename: 'screenshots.html',
        pathBuilder: function(currentSpec) {
          return browser.capabilities.get('platform') + '/' + browser.capabilities.get('browserName') + '.' + browser.capabilities.get('version') + '/' + currentSpec.fullName;
        }
      })
    );
  },

```

**Comment by ammachado** [comment]
+1

**Comment by AndrewFinlay** [comment]
I've had a look at this and think that this issue may be getting a little confused.  Both the original method and @bvarga 's method of getting browser capabilities into the pathBuilder seem to work fine.  

The problem that the original modification is trying to solve is that sometimes screenshots aren't captured if the failing test is the last spec.  This seems to be due to something deeper in the Protractor stack, that fails to wait for the async finish of specDone.  After running a few tests I've found that none of the changes here prevent that.

Why not go back to the original implementation to restore the browser.capability support, then as a workaround set your last spec to:

``` javascript
   it('This empty spec is a hack fix for Jasmine Reporters bug.', function() {});
```

or from the linked thread, add this to a containing suite

``` javascript
afterAll(function(done){
    process.nextTick(done);
});
```

**Comment by bmsoko** [comment]
Can you please share what's the work around with this?

Thanks!!

**Comment by mlison** [comment]
closed by #29

**Comment by kummerer94** [comment]
I'm using Protractor 3.1 and the 0.3.0 version of protractor-jasmine-2-screenshot-reporter and still see this problem.

```
[INFO] [firefox #2] T:/codebase/code/ui/node_modules/protractor/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:1337
[INFO] [firefox #2]       return callbackFn(this.value_);
[INFO] [firefox #2]              ^
[INFO] [firefox #2] TypeError: Cannot read property 'getCapabilities' of undefined
[INFO] [firefox #2]     at T:\codebase\code\ui\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver\webdriver.js:432:19
[INFO] [firefox #2]     at promise.Promise.goog.defineClass.invokeCallback_ (T:/codebase/code/ui/node_modules/protractor/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:1337:14)
[INFO] [firefox #2]     at promise.ControlFlow.goog.defineClass.goog.defineClass.abort_.error.executeNext_.execute_ (T:/codebase/code/ui/node_modules/protractor/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:2776:14)
[INFO] [firefox #2]     at promise.ControlFlow.goog.defineClass.goog.defineClass.abort_.error.executeNext_ (T:/codebase/code/ui/node_modules/protractor/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:2758:21)
[INFO] [firefox #2]     at goog.async.run.processWorkQueue (T:\codebase\code\ui\node_modules\protractor\node_modules\selenium-webdriver\lib\goog\async\run.js:124:15)
[INFO] [firefox #2]     at process._tickCallback (node.js:368:9)
```

Does anyone know a fix for this?
